paparazzi/src/main/java/app/cash/paparazzi/internal/resources/PseudolocaleGenerator.kt

```diff
@@ -0,0 +1,144 @@
+import app.cash.paparazzi.internal.resources.base.BasicPluralsResourceItem
+import app.cash.paparazzi.internal.resources.base.BasicTextValueResourceItem
+import app.cash.paparazzi.internal.resources.base.BasicValueResourceItem
+import app.cash.paparazzi.internal.resources.base.BasicValueResourceItemBase
+import com.android.ide.common.rendering.api.ResourceNamespace
+import com.android.ide.common.resources.ResourceValueMap
+import com.android.ide.common.resources.configuration.FolderConfiguration
+import com.android.ide.common.resources.configuration.LocaleQualifier
+import com.android.resources.ResourceType
+import com.google.common.collect.Table
+
+private const val DO_NOT_TRANSLATE = "donottranslate"
+
+private val configCache = hashMapOf<Pair<ResourceSourceFile, LocaleQualifier>, RepositoryConfiguration>()
+
+fun String.pseudoLocalize(method: Pseudolocalizer.Method) =
+  if (isBlank()) {
+    this
+  } else {
+    with(Pseudolocalizer(method)) {
+      start() + text(this@pseudoLocalize) + end()
+    }
+  }
+
+fun Table<ResourceNamespace, ResourceType, ResourceValueMap>.pseudolocalizeIfNeeded(
+  localeQualifier: LocaleQualifier
+): Table<ResourceNamespace, ResourceType, ResourceValueMap> {
+  for (namespace in rowKeySet()) {
+    for (type in columnKeySet()) {
+      val map = this[namespace, type]
+      map?.forEach { (name, value) ->
+        map[name] =
+          (value as? BasicValueResourceItemBase)?.pseudolocalizeIfNeeded(localeQualifier) ?: value
+      }
+    }
+  }
+
+  return this
+}
+
+private fun BasicValueResourceItemBase.pseudolocalizeIfNeeded(
+  localeQualifier: LocaleQualifier
+): BasicValueResourceItemBase {
+  if (!isPseudolocalizable(this)) {
+    return this
+  }
+
+  val method = when (localeQualifier.value) {
+    "en-rXA" -> Pseudolocalizer.Method.ACCENT
+    "ar-rXB" -> Pseudolocalizer.Method.BIDI
+    else -> return this
+  }
+
+  var repositoryConfiguration = configCache[sourceFile to localeQualifier]
+  if (repositoryConfiguration == null) {
+    repositoryConfiguration = sourceFile.configuration.onNewLocaleQualifier(localeQualifier)
+    configCache[sourceFile to localeQualifier] = repositoryConfiguration
+  }
+
+  val pseudoLocaleSourceFile = sourceFile.onNewRepositoryConfiguration(repositoryConfiguration)
+  val pseudoItem: BasicValueResourceItemBase = when (resourceType) {
+    ResourceType.STRING -> pseudolocalizeString(
+      this as BasicValueResourceItem,
+      pseudoLocaleSourceFile,
+      method
+    )
+
+    ResourceType.PLURALS -> pseudolocalizePlural(
+      this as BasicPluralsResourceItem,
+      pseudoLocaleSourceFile,
+      method
+    )
+
+    ResourceType.ARRAY -> pseudolocalizeArray(
+      this as BasicArrayResourceItem,
+      pseudoLocaleSourceFile,
+      method
+    )
+
+    else -> return this
+  }
+
+  return pseudoItem
+}
+
+private fun isPseudolocalizable(resourceItemBase: BasicValueResourceItemBase): Boolean =
```

Comment:
should this also check for:
1) whether this is the default locale (i.e., only pseudolocalize strings that are in the default locale)
2) whether the string is translatable, i.e., `translatable="false"`?

This would align it more to AAPT's pseudolocalizable check: https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:build-system/aaptcompiler/src/main/java/com/android/aaptcompiler/PseudolocaleGenerator.kt;l=217?q=isPseudolo&sq=